src/general_listener.c

```diff
@@ -129,103 +129,191 @@ static void
 peak_general_listener_on_enter(GumInvocationListener* listener,
                                GumInvocationContext* ic)
 {
-    if (!listener || g_object_is_floating(listener)) {
-        return;
-    }
-    PeakGeneralListener* self = PEAKGENERAL_LISTENER(listener);
-    pthread_t my_tid = pthread_self();
-    size_t mapped_tid = (size_t)(gum_metal_hash_table_lookup(peak_tid_mapping, GUINT_TO_POINTER(my_tid)));
-    // g_print ("hook_id %lu tid %lu mapped %lu\n", hook_id, pthread_self(), mapped_tid);
-    // g_print ("hook_id %lu max %lu tid %lu ncall %p \n", hook_id, peak_max_num_threads, mapped_tid, self->num_calls);
-    size_t index = mapped_tid;
-    if (thread_data.child_time == NULL) {
-        thread_data.level = 0;
-        thread_data.capacity = 16;
-        pthread_pause_disable();
-        thread_data.child_time = g_new(gdouble, 16);
-        pthread_pause_enable();
-    }
-    thread_data.child_time[thread_data.level] = 0.0;
-    thread_data.level++;
-    if (thread_data.level == thread_data.capacity) {
-        thread_data.capacity *= 2;
-        pthread_pause_disable();
-        thread_data.child_time = g_renew(double, thread_data.child_time, thread_data.capacity);
-        pthread_pause_enable();
-    }
-    self->num_calls[index]++;
-    if (peak_detach_count && self->num_calls[index] >= peak_detach_count) {
-        pthread_mutex_lock(&lock);
-        size_t hook_id = self->hook_id;
-        if (!array_listener_detached[hook_id]) {
-            array_listener_detached[hook_id] = TRUE;
-            GumMetalHashTableIter peak_tid_iter;
-            pthread_t peak_tid_key;
-            gum_metal_hash_table_iter_init(&peak_tid_iter, peak_tid_mapping);
-            while (gum_metal_hash_table_iter_next(&peak_tid_iter, (void**)&peak_tid_key, NULL)) {
-                // g_print ("peak_tid_key %lu my_tid %lu\n", peak_tid_key, my_tid);
-                if (peak_tid_key != my_tid)
-                    pthread_pause(peak_tid_key);
+    if (peak_detach_count == 0) {
+        if (!listener || g_object_is_floating(listener)) {
+            return;
+        }
+        PeakGeneralListener* self = PEAKGENERAL_LISTENER(listener);
+        pthread_t my_tid = pthread_self();
+        size_t mapped_tid = (size_t)(gum_metal_hash_table_lookup(peak_tid_mapping, GUINT_TO_POINTER(my_tid)));
```

Comment:
These lines should be move outside of the if else statement here to reduce code duplication.